include/robot_interfaces/n_joint_robot_functions.hpp

```diff
@@ -0,0 +1,131 @@
+// BSD 3-Clause License
+//
+// Copyright (C) 2018-2019, Max Planck Gesellschaft
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "n_joint_robot_types.hpp"
+
+namespace robot_interfaces
+{
+/**
+ * @brief Collection of functions for a generic N-joint BLMC robot.
+ *
+ * Defines all the types needed to set up an interface to a generic N-joint BLMC
+ * robot that expects as Action a simple vector of N torque commands and
+ * provides N observations containing measured joint angle, velocity and torque.
+ *
+ * Defines some generic functions that can be used by all robots that are based
+ * on the `NJointRobotTypes`.
+ *
+ * @tparam N_JOINTS Number of joints the robot has.
+ */
+template <size_t N_JOINTS>
+struct NJointRobotFunctions
+{
+    using Types = NJointRobotTypes<N_JOINTS>;
+
+    /**
+     * @brief Process the desired action provided by the user.
+     *
+     * Takes the desired action from the user and does the following processing:
+     *
+     * ## 1. Run the position controller in case a target position is set.
+     *
+     *   If the target position is set to a value unequal to NaN for any joint,
+     *   a PD position controller is executed for this joint and the resulting
+     *   torque command is added to the torque command in the action.
+     *
+     *   If the P- and/or D-gains are set to a non-NaN value in the action, they
+     *   are used for the control.  NaN-values are replaced with the default
+     *   gains.
+     *
+     * ## 2. Apply safety checks.
+     *
+     *   - Limit the torque to the allowed maximum value.
+     *   - Dampen velocity using the given safety_kd gains.  Damping us done
+     *     joint-wise using this equation:
+     *
+     *         torque_damped = torque_desired - safety_kd * current_velocity
+     *
+     * The resulting action with modifications of all steps is returned.
+     *
+     * @param desired_action  Desired action given by the user.
+     * @param latest_observation  Latest observation from the robot.
+     * @param max_torque_Nm  Maximum allowed absolute torque.
+     * @param safety_kd  D-gain for velocity damping.
+     * @param default_position_control_kp  Default P-gain for position control.
+     * @param default_position_control_kd  Default D-gain for position control.
+     *
+     * @return Resulting action after applying all the processing.
+     */
+    static typename Types::Action process_desired_action(
+        const typename Types::Action &desired_action,
+        const typename Types::Observation &latest_observation,
+        const double max_torque_Nm,
+        const typename Types::Vector &safety_kd,
+        const typename Types::Vector &default_position_control_kp,
+        const typename Types::Vector &default_position_control_kd)
```

Comment:
I see that this is a bit dirty but I am not sure how to solve this. The only clean solution that I can see at the moment is to use different types for desired and applied action but this would add too much complexity to the interface in my opinion.